File: day10/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>
#include <sstream>
#include <unordered_set>
#include <unordered_map>
#include <map>

auto read() {
    std::fstream ifs("/Users/ecem/CLionProjects/adventofcode2022/day10/day10.txt");
    std::string line;
    std::vector<std::string> result;
    while (std::getline(ifs, line)) {
        if (line.empty())
            continue;
        result.push_back(line);
    }
    return result;
}

std::vector<std::string> split(const std::string &str) {
    std::istringstream split(str);
    std::vector<std::string> tokens;
    for (std::string each; std::getline(split, each, ' '); tokens.push_back(each));
    return tokens;
}

void part1() {

    int64_t result = 0;
    int64_t x = 1;
    int64_t cycle = 0;
    int64_t next_cycle = 20;

    for (const auto &line: read()) {
        auto tokens = split(line);

        if (tokens[0] == "addx") {
            int64_t value = std::stoi(tokens[1]);
            cycle += 2;

            if (next_cycle <= cycle) {
                result += (next_cycle * x);
                next_cycle += 40;
            }

            x += value;
        } else {
            assert(tokens[0] == "noop");
            cycle += 1;

            if (next_cycle <= cycle) {
                result += (next_cycle * x);
                next_cycle += 40;
            }
        }
    }

    std::cout << "Part1: " << result << std::endl;
}

void check(int64_t x, int64_t &cycle, std::vector<char> &crt) {
    if (x - 1 <= (cycle % 40) && (cycle % 40) <= x + 1) {
        crt[cycle] = '#';
    }

    ++cycle;
}

void part2() {

    int64_t x = 1;
    int64_t cycle = 0;
    std::vector<char> crt = std::vector<char>(240, ' ');

    for (const auto &line: read()) {
        auto tokens = split(line);

        if (tokens[0] == "addx") {
            int64_t value = std::stoi(tokens[1]);
            check(x, cycle, crt);
            check(x, cycle, crt);
            x += value;
        } else {
            assert(tokens[0] == "noop");
            check(x, cycle, crt);
        }
    }

    std::cout << "Part2:" << std::endl;
    for (int i = 0; i + 40 <= crt.size(); i += 40) {
        std::cout << std::string(crt.begin() + i, crt.begin() + i + 40) << std::endl;
    }
}

int main() {
    part1();
    part2();
    return 0;
}

File: day10/src/main.rs
use std::cell::RefCell;
use std::fs::File;
use std::io::{self, BufRead};
use std::ops::Deref;
use std::path::Path;
use std::rc::Rc;
use std::vec;
use scanf::sscanf;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>> where P: AsRef<Path> {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "/Users/ecem/CLionProjects/adventofcode2022/day10/day10.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}


fn part1() {
    let mut result = 0;
    let mut x = 1;
    let mut cycle = 0;
    let mut next_cycle = 20;

    for line in read() {
        let tokens = line.split(' ').collect::<Vec<&str>>();

        match tokens[0] {
            "addx" => {
                let value = tokens[1].parse::<i32>().unwrap();
                cycle += 2;
                if next_cycle <= cycle {
                    result += next_cycle * x;
                    next_cycle += 40;
                }

                x += value;
            }
            "noop" => {
                cycle += 1;
                if next_cycle <= cycle {
                    result += next_cycle * x;
                    next_cycle += 40;
                }
            }
            op => panic!("unknown command {}", op)
        }
    }

    println!("Part1: {}", result)
}

fn check(x: i32, cycle: &mut i32, crt: &mut Vec<char>) {
    let c = *cycle;
    if x - 1 <= (c % 40) && (c % 40) <= x + 1 {
        crt[c as usize] = '#';
    }
    *cycle += 1;
}

fn part2() {
    let mut x = 1;
    let mut cycle = 0;
    let mut crt = vec![' '; 240];

    for line in read() {
        let tokens = line.split(' ').collect::<Vec<&str>>();

        match tokens[0] {
            "addx" => {
                let value = tokens[1].parse::<i32>().unwrap();
                check(x, &mut cycle, &mut crt);
                check(x, &mut cycle, &mut crt);
                x += value;
            }
            "noop" => {
                check(x, &mut cycle, &mut crt);
            }
            op => panic!("unknown command {}", op)
        }
    }

    println!("Part2:");
    for chunk in crt.chunks(40) {
        println!("{}", chunk.iter().collect::<String>())
    }
}

fn main() {
    part1();
    part2();
}


